tests/gradients/test_finite_difference.py

```diff
@@ -0,0 +1,590 @@
+        assert np.allclose(coeffs, [-1.5, 2, -0.5])
+        assert np.allclose(shifts, [0, 1, 2])
+
+    def test_correct_center_order2(self):
+        """Test that the correct centered order 2 method is returned"""
+        coeffs, shifts = finite_diff_stencil(1, 2, "center")
+        assert np.allclose(coeffs, [-0.5, 0.5])
+        assert np.allclose(shifts, [-1, 1])
+
+    def test_correct_backward_order1(self):
+        """Test that the correct backward order 1 method is returned"""
+        coeffs, shifts = finite_diff_stencil(1, 1, "backward")
+        assert np.allclose(coeffs, [1, -1])
+        assert np.allclose(shifts, [0, -1])
+
+    def test_correct_second_derivative_forward_order1(self):
+        """Test that the correct forward order 1 method is returned"""
+        coeffs, shifts = finite_diff_stencil(2, 1, "forward")
+        assert np.allclose(coeffs, [1, -2, 1])
+        assert np.allclose(shifts, [0, 1, 2])
+
+    def test_correct_second_derivative_center_order4(self):
+        """Test that the correct forward order 1 method is returned"""
+        coeffs, shifts = finite_diff_stencil(2, 4, "center")
+        assert np.allclose(coeffs, [-2.5, 4 / 3, 4 / 3, -1 / 12, -1 / 12])
+        assert np.allclose(shifts, [0, -1, 1, -2, 2])
+
+
+class TestShiftedTapes:
+    """Tests for the generate_shifted_tapes function"""
+
+    def test_behaviour(self):
+        """Test that the function behaves as expected"""
+
+        with qml.tape.QuantumTape() as tape:
+            qml.PauliZ(0)
+            qml.RX(1.0, wires=0)
+            qml.CNOT(wires=[0, 2])
+            qml.Rot(2.0, 3.0, 4.0, wires=0)
+            qml.expval(qml.PauliZ(0))
+
+        tape.trainable_params = {0, 2}
+        shifts = [0.1, -0.2, 1.6]
+        res = generate_shifted_tapes(tape, 1, shifts=shifts)
+
+        assert len(res) == len(shifts)
+        assert res[0].get_parameters(trainable_only=False) == [1.0, 2.0, 3.1, 4.0]
+        assert res[1].get_parameters(trainable_only=False) == [1.0, 2.0, 2.8, 4.0]
+        assert res[2].get_parameters(trainable_only=False) == [1.0, 2.0, 4.6, 4.0]
+
+
+class TestFiniteDiff:
+    """Tests for the finite difference gradient transform"""
+
+    def test_non_differentiable_error(self):
+        """Test error raised if attempting to differentiate with
+        respect to a non-differentiable argument"""
+        psi = np.array([1, 0, 1, 0]) / np.sqrt(2)
+
+        with qml.tape.JacobianTape() as tape:
+            qml.QubitStateVector(psi, wires=[0, 1])
+            qml.RX(0.543, wires=[0])
+            qml.RY(-0.654, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            qml.probs(wires=[0, 1])
+
+        # by default all parameters are assumed to be trainable
+        with pytest.raises(
+            ValueError, match=r"Cannot differentiate with respect to parameter\(s\) {0}"
+        ):
+            finite_diff(tape)
+
+        # setting trainable parameters avoids this
+        tape.trainable_params = {1, 2}

```

Comment:
I'm a bit confused, if these are just explicit floats how can they be set as trainable parameters?

Reply:
Oh, the issue is `psi` for the `QubitStateVector`! Parameter number 0 corresponds to this parameter

Reply:
I think @glassnotes question may have been how the floats in `qml.RX(0.543, wires=[0])` and `qml.RY(-0.654, wires=[1])` can be trainable if we did not use autograd's numpy?

If so, I think the answer is that the finite diff method applies in the "blackbox" where no interface tracking is happening any more, and when all parameters are just numpy objects or floats anyways. (If not ignore :) )

Reply:
Oh, in that case - this is because PL currently assumes floats are trainable by default (and will, behind the scenes, do `np.array(x, requires_grad=True)`. This is behaviour I would like to change soon :)

Reply:
Ahh okay gotcha :+1: Yeah I didn't realize that change happens under the hood, that's why I was confused.